unittests/savanna_disaster_recovery.cpp

```diff
@@ -0,0 +1,270 @@
+#include "savanna_cluster.hpp"
+
+using namespace eosio::chain;
+using namespace eosio::testing;
+
+BOOST_AUTO_TEST_SUITE(savanna_disaster_recovery)
+
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(node_goes_down, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0]; auto& C=_nodes[2];
+
+   C.close();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(4);  }), 4); // lib still advances with 3 finalizers
+   C.open();
+   A.push_blocks_to(C);
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(4);  }), 4); // all 4 finalizers should be back voting
+   BOOST_REQUIRE(!C.is_head_missing_finalizer_votes());                        // let's make sure of that
+} FC_LOG_AND_RETHROW()
+
+
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(recover_killed_node_with_old_fsi, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0]; auto& C=_nodes[2];
+
+   auto fsi = C.save_fsi();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   auto snapshot = C.snapshot();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   C.close();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // lib still advances with 3 finalizers
+   C.remove_state();
+   C.overwrite_fsi(fsi);
+   C.open_from_snapshot(snapshot);
+   A.push_blocks_to(C);
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // all 4 finalizers should be back voting
+   BOOST_REQUIRE(!C.is_head_missing_finalizer_votes());                        // let's make sure of that
+} FC_LOG_AND_RETHROW()
+
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(recover_killed_node_with_deleted_fsi, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0]; auto& C=_nodes[2];
+
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   auto snapshot = C.snapshot();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   C.close();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // lib still advances with 3 finalizers
+   C.remove_state();
+   C.remove_fsi();
+   C.open_from_snapshot(snapshot);
+   A.push_blocks_to(C);
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // all 4 finalizers should be back voting
+   BOOST_REQUIRE(!C.is_head_missing_finalizer_votes());                        // let's make sure of that
+} FC_LOG_AND_RETHROW()
+
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(recover_killed_node_while_retaining_fsi, savanna_cluster::cluster_t) try {
```

Comment:
Is this the same as `recover_killed_node_with_old_fsi`? The two tests are using the old FSI.

Reply:
No, this one uses the fsi that has always been kept up to date with all the votes.
`recover_killed_node_with_old_fsi` saves a fsi, has the node vote advancing its lock (so the live fsi is different from the saved one), and then kills the node and restarts it with the old fsi.